### ISSUE TITLE
Feature request: allow transformations between descriptor.proto objects and reflection objects

### ISSUE THREAD

**Issue by murgatroid99** [dcodeIO/protobuf.js#757]
protobuf.js version: 6

It would be useful to be able to take a message of a type defined in `google/protobuf/descriptor.proto`, and convert it into the corresponding reflection object, and also be able to convert the other way. Of course, some of them don't really make sense, but I would like, for example, to be able to `load` a `DescriptorProto` message instance as a `Message` object, and conversely to be able to extract a `DescriptorProto` instance from a `Message` object.

**Comment by dcodeIO** [comment]
That's basically [self-describing messages](https://developers.google.com/protocol-buffers/docs/techniques#self-description)-support, right?

**Comment by murgatroid99** [comment]
Not really. The types in `descriptor.proto` are structurally equivalent to the information you can load from a `.proto` file. For example, and instance of the `DescriptorProto` message type contains the same information as a `message` declaration in a `.proto` file. What I'm asking for is a transformation between instances of these messages and the ProtoBuf.js reflection types.

**Comment by dcodeIO** [comment]
To my understanding, `DescriptorProto`s reference enums and messages of other `DescriptorProto`s, including only nested enums and messages. Hence, resolving any such type referncing a non-nested enum or message would throw.

Only a `FileDescriptorSet`, which is used in self-describing messages, would contain all relevant `DescriptorProto`s making it possibly to successfully work with (like encode and decode) the types imported this way.

I might just be misunderstanding your use-case, though, so feel free to point me into the right direction.

**Comment by murgatroid99** [comment]
Right, `DescriptorProto` may have been a bad example. It is likely that I would only need to transform between instances of `FileDescriptorSet` and `Root` objects. The goal here is to support the [gRPC server reflection protocol](https://github.com/grpc/grpc/blob/master/src/proto/grpc/reflection/v1alpha/reflection.proto).

**Comment by dcodeIO** [comment]
I quickly put something together as a starting point, see commit above. The reason why I'd prefer an extension here is that it's usually not necessary to include the entire descriptor (functionality) as part of the standard library.

Idea is that doing `require("protobufjs/ext/descriptor")` loads the extension along with the descriptor definitions and then adds `fromDescriptor`, `#toDescriptor` methods to all reflection types. See `protobuf.Type.fromDescriptor` for how these methods could look like.

**Comment by murgatroid99** [comment]
That looks like what I was asking for.

**Comment by dcodeIO** [comment]
Sigh, this will take a while. Added a few comments to document what's (still) necessary.

**Comment by dcodeIO** [comment]
Still pretty rough around the edges, but it's now able to convert a root with loaded descriptor.proto in it to a FileDescriptorSet, encode it, decode it back and have it return something similar (still needs a real test suite) to the descriptor we initially put in. Wheee...

**Comment by dcodeIO** [comment]
With `ext/descriptor` nowadays maintaining enough metadata for the `protoc-gen-pbjs` plugin to pass the conformance harness, we can proabably consider this issue addressed.
